sopel/formatting.py

```diff
@@ -247,3 +310,12 @@ def reverse(text):
         that understand it (e.g. mIRC) can be unpredictable. Use it carefully.
     """
     return ''.join([CONTROL_REVERSE, text, CONTROL_REVERSE])
+
+
+def plain(text):
+    """Return the text without any IRC formatting.
+
+    :param str text: text with potential IRC formatting control code(s)
+    :rtype: str
+    """
+    return PLAIN_REGEX.sub('', text)
```

Comment:
Since I was thinking of making at least part of this PR myself, I looked at [ZNC's implementation](https://github.com/znc/znc/blob/700fdab3fb61061dd1b6ae46a191e040cc434c8d/src/ZNCString.cpp#L1393-L1440) a while back. Could we maybe use an iterative approach instead of regex? Since this is going to be called on every message Sopel receives, it's worth thinking about perf.

My simplistic mockup (it's past midnight, after all 😛) of a "dumb" `plain()` function using iteration did pretty well against the regex-based implementation you included here:

```
>>> text_text = "This is a string with \x02control\x02 \x1dcodes\x1d in it."
>>> from timeit import timeit
>>> from sopel.formatting import plain
>>> def plain_dumb(text):
...     out = []
...     for c in text:
...         if ord(c) < 0x20:
...             continue
...         out.append(c)
...     return ''.join(out)
... 
>>> assert plain(test_text) == plain_dumb(test_text)
>>> timeit(stmt='plain(test_text)', globals=globals())
7.396932500007097
>>> timeit(stmt='plain_dumb(test_text)', globals=globals())
3.493459499964956
```

Obviously my `plain_dumb()` function needs a little additional state-machine logic to account for formatting codes that take parameters, but even with that I bet it'd still be a large net perf gain vs. using regex.

(Commenting here because the infra above for this implementation is really long, but the actual function is just one line of "real code". 😝)

Reply:
**TL;DR**: Will try. It's way funnier to do the iterative approach, the regex is my safety net, because at Sopel's scale, the difference is nothing.

Long version: you got a fair point, of a function ~2.1 times faster than the regex version. However, since `timeit` runs the code about a million times, if we scale down to what we can expect with an IRC server - say, about 1000 messages per minutes at best? - we can argue that the overhead of the `plain` function is close to 7.4ms per minute with regex, 3.5ms per minute with iterative approach. So if I can't pull the iterative approach, we are fine. I'll definitively try it, because it's way more fun, but I'm not worried, and I hope you are not either.

Since all the tests are written independently from the implementation, I can safely try. So: hurray! \o/

Reply:
After a bit of work I came up with e35457574f6ebca43e4709b70940d26a4923cba5 which is a trade-off: I couldn't get a proper strip function with the (hex) color control codes, it was all messy and eventually more time consuming (yeah, perf!).

So here it is, if there is a (hex) color control code, it uses a regex. Otherwise, it doesn't. Works fine, tests are all green, so I'm pretty happy about it.

You can test the perf, it's way better, but not as good as a ~3.49s for 1,000,000: more like 4.5 without colors, 5.5 with colors. Not bad at all!